From eadadc2c847ee55dcd5fa7f5c636874c8a63abd9 Mon Sep 17 00:00:00 2001
Message-ID: <eadadc2c847ee55dcd5fa7f5c636874c8a63abd9.1779975966.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1779975966.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Fri, 22 May 2026 11:21:33 +0200
Subject: [PATCH 11/30] virtio-msg-bus: Report disconnected peers as retryable

A disconnected virtio-msg bus peer is a transport readiness condition, not
a successfully delivered message.

Return VIRTIO_MSG_ERROR_RETRY when send or receive paths are used before
the peer is connected so callers can preserve required traffic or retry
after the transport becomes ready.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-msg-bus.c         |  4 ++++
 include/hw/virtio/virtio-msg-bus.h | 15 ++++++++++++---
 2 files changed, 16 insertions(+), 3 deletions(-)

diff --git a/hw/virtio/virtio-msg-bus.c b/hw/virtio/virtio-msg-bus.c
index e907fd64ecce..abf2218d9b21 100644
--- a/hw/virtio/virtio-msg-bus.c
+++ b/hw/virtio/virtio-msg-bus.c
@@ -41,6 +41,10 @@ int virtio_msg_bus_send(BusState *bus, VirtIOMSG *msg_req)
     VirtIOMSGBusDeviceClass *bdc;
     int r = VIRTIO_MSG_NO_ERROR;
 
+    if (!virtio_msg_bus_peer_ready(bd)) {
+        return VIRTIO_MSG_ERROR_RETRY;
+    }
+
     bdc = VIRTIO_MSG_BUS_DEVICE_CLASS(object_get_class(OBJECT(bd)));
 
     if (bdc->send) {
diff --git a/include/hw/virtio/virtio-msg-bus.h b/include/hw/virtio/virtio-msg-bus.h
index c01681824c25..d99d5369aea9 100644
--- a/include/hw/virtio/virtio-msg-bus.h
+++ b/include/hw/virtio/virtio-msg-bus.h
@@ -66,11 +66,16 @@ static inline VirtIOMSGBusDevice *virtio_msg_bus_get_device(BusState *qbus)
     return (VirtIOMSGBusDevice *)qdev;
 }
 
+static inline bool virtio_msg_bus_peer_ready(VirtIOMSGBusDevice *bd)
+{
+    return bd && bd->peer != NULL;
+}
+
 static inline bool virtio_msg_bus_connected(BusState *bus)
 {
     VirtIOMSGBusDevice *bd = virtio_msg_bus_get_device(bus);
 
-    return bd && bd->peer != NULL;
+    return virtio_msg_bus_peer_ready(bd);
 }
 
 void virtio_msg_bus_process(VirtIOMSGBusDevice *bd);
@@ -79,11 +84,15 @@ bool virtio_msg_bus_connect(BusState *bus,
                             const VirtIOMSGBusPort *port,
                             void *opaque);
 
-static inline void
+static inline int
 virtio_msg_bus_receive(VirtIOMSGBusDevice *bd, VirtIOMSG *msg)
 {
+    if (!virtio_msg_bus_peer_ready(bd)) {
+        return VIRTIO_MSG_ERROR_RETRY;
+    }
+
     virtio_msg_unpack(msg);
-    bd->peer->receive(bd, msg);
+    return bd->peer->receive(bd, msg);
 }
 
 int virtio_msg_bus_send(BusState *bus, VirtIOMSG *msg_req);
-- 
2.53.0

